ops/model.py

```diff
@@ -349,6 +358,7 @@ def __init__(self, meta: _charm.CharmMeta, backend: _ModelBackend):
         self._secret_set_cache: collections.defaultdict[str, dict[str, Any]] = (
             collections.defaultdict(dict)
         )
+        self._databag_obj_cache: dict[tuple[int, Union[Application, Unit]], Any] = {}
```

Comment:
Now that the future annotations are in this file, this can be:

```suggestion
        self._databag_obj_cache: dict[tuple[int, Application | Unit], Any] = {}
```